staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go

```diff
@@ -540,6 +543,53 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
 			return err
 		}
 
+		// Check if the returned object needs to have its
+		// kubectl-client-side-apply managed fields migrated server-side-apply.
+		//
+		// By checking afterward instead of fetching the object beforehand and
+		// unconditionally fetching we can make 3 network requests in the rare
+		// case of migration and 1 request if migration is unnecessary.
+		//
+		// To check beforehand means 2 requests for most operations, and 3
+		//	requests in worst case.
+		if patchData, err := o.calculateSSAMigrationPatch(obj); err != nil {
```

Comment:
This approach differs a big from the approach discussed in the SIG. During the SIG, we proposed for SSA to unconditionally fetch the object and patch the fields BEFORE the apply is done.

I propose this alternative method which preserve's SSA's property of only needing 1 request in the average case. But it does complicate the error handling quite a bit.

Thoughts?

Reply:
This is a better approach. No extra retrievals for the common (no migration) case is an excellent approach.